### DISCUSSION TITLE
Do not activate by default if the Editor translation is not 100%.

### DISCUSSION THREAD

**Discussion by setanarut** [godotengine/godot-proposals#4327] [Editor]
Do not activate by default if the translation is not 100%. When the first installation is done, an interface with two languages mixed is not nice at all.

# Proposal

1. If the translation is not 100% make it the default English.
2. Add the percentage of each translation to the menu below.  

<img width="568" alt="langs" src="https://user-images.githubusercontent.com/36481442/161305485-2912b634-846b-4928-ad4d-53c06e4f35ab.png">
.

**Comment by Mickeon** [comment]
Let's be honest, complete 100% is not easily achievable, especially as features are added or tweaked around. Even at around 98%, it's very likely the translated editor is usable without any major slip-ups.

**Reply by setanarut** [reply]
It could be 98%. maybe 96-97%. I think these ratios are acceptable.

**Reply by Cerno-b** [reply]
Looking at the current state of Weblate, maybe 95% is a good cutoff point, there seems to be a drop after that even though it's just a momentary observation.

![image](https://github.com/godotengine/godot-proposals/assets/16470072/b23cd09c-0f5e-492c-b3f2-ce6cd52d35ac)

**Comment by Calinou** [comment]
I don't think a cutoff at < 100% completion is a good idea. It's too strict and will often end up not being reached anymore following a Godot update.

What I'd do instead is list the completion percentage next to each translation in the language selection dropdown.

[Incomplete translations are already excluded from the editor binary](https://github.com/godotengine/godot/pull/54020). As of writing, the completion cutoffs are < 30% (editor) and < 10% (class reference). We can make the cutoff stricter as time goes on.

**Reply by setanarut** [reply]
%30 is too low. Issue is Mixed languages.

**Reply by SilencedPerson** [reply]
Mixed language is still better than no language.

**Reply by setanarut** [reply]
> Mixed language is still better than no language.

Bullshit. Mixed language is worse than just English.

**Reply by SilencedPerson** [reply]
How is more English useful for someone who doesn't speak English?

**Reply by setanarut** [reply]
You can choose from the menu as 30% native language and 70% English. It's unusual to impose this by default.

**Reply by Cerno-b** [reply]
If this issue gets realized, will there still be the need to strictly exclude languages with a low translation rate? 

I understand that it makes sense to not show them by default, but someone who really struggles with English might be happy to have the option to at least understand a few crumbs instead of being completely lost. Also displaying the translation percentage prepares the user for a potentially bumpy ride and it might motivate people to contribute to translations. 

Or is the reason to exclude the languages to limit the length of the list?

**Comment by Cerno-b** [comment]
@Calinou Here's a snippet that works for me:

```python
import re

# pick any of the the three
path = r"[godot_docs_l10n_repo]\weblate\de.po"
path = r"[godot_repo]\editor\translations\editor\de"
path = r"[godot_repo]\editor\translations\properties\de"

with open(path, encoding="utf-8") as f:
    text = f.read()

# Total count and untranslated count could be off by 1 but the precision is not needed
# and it makes for more readable code.

# Count all occurrences of 'msgid' at the start of a line.
total_count = len(re.findall(r'[\r\n]msgid', text))

# Count all occurrences of 'msgstr ""' at the start of a line 
# with any trailing whitespace or linebreaks.
# The hash character marks the next translation string. 
# This is necessary because translated strings sometimes take
# the form 'msgstr ""' followed by the actual translation in the next line.
untranslated_count = len(re.findall(r'[\r\n]msgstr ""[\r\n\w]*#', text))

# Count all occurrences of the fuzzy marker
fuzzy_count = len(re.findall(r'[\r\n]#, fuzzy', text))

translated_percentage = 1 - (untranslated_count + fuzzy_count) / total_count
translated_percentage_int = int(100*translated_percentage)

# This should print the percentage that weblate shows for de.
print(f"{translated_percentage_int}%")
```

**Comment by setanarut** [comment]
> @Calinou Here's a snippet that works for me:
> 
> ```python
> import re
> 
> # pick any of the the three
> path = r"[godot_docs_l10n_repo]\weblate\de.po"
> path = r"[godot_repo]\editor\translations\editor\de"
> path = r"[godot_repo]\editor\translations\properties\de"
> 
> with open(path, encoding="utf-8") as f:
>     text = f.read()
> 
> # Total count and untranslated count could be off by 1 but the precision is not needed
> # and it makes for more readable code.
> 
> # Count all occurrences of 'msgid' at the start of a line.
> total_count = len(re.findall(r'[\r\n]msgid', text))
> 
> # Count all occurrences of 'msgstr ""' at the start of a line 
> # with any trailing whitespace or linebreaks.
> # The hash character marks the next translation string. 
> # This is necessary because translated strings sometimes take
> # the form 'msgstr ""' followed by the actual translation in the next line.
> untranslated_count = len(re.findall(r'[\r\n]msgstr ""[\r\n\w]*#', text))
> 
> # Count all occurrences of the fuzzy marker
> fuzzy_count = len(re.findall(r'[\r\n]#, fuzzy', text))
> 
> translated_percentage = 1 - (untranslated_count + fuzzy_count) / total_count
> translated_percentage_int = int(100*translated_percentage)
> 
> # This should print the percentage that weblate shows for de.
> print(f"{translated_percentage_int}%")
> ```

They will acknowledge the problem after 3 years, don't waste your time.

**Reply by Mickeon** [reply]
The reason why this snippet of code was sent is because it was in fact [acknowledged on RocketChat](https://chat.godotengine.org/channel/translation?msg=4Nr94SpggehfjjbKf) and theorized that showing the completion percentage next to the available translations should be possible.

It does not address this original issue fully, but it should make it easier for users to infer when the translation is good enough, for now.

**Reply by setanarut** [reply]
> The reason why this snippet of code was sent is because it was in fact [acknowledged on RocketChat](https://chat.godotengine.org/channel/translation?msg=4Nr94SpggehfjjbKf) and theorized that showing the completion percentage next to the available translations should be possible.
> 
> It does not address this original issue fully, but it should make it easier for users to infer when the translation is good enough, for now.

2 years ago I opened this thread and yesterday I closed it and when I closed it they became interested. that's why. And he's still talking about 50%.

**Reply by Mickeon** [reply]
> 2 years ago I opened this thread and yesterday I closed it and when I closed it they became interested.

Sometimes all it takes is a spark. When the discussion was closed, I believe either Calinou received a notification, or it brought the discussion above the ones with most recent activity. There's a lot of discussions of various topics. It's difficult to keep track of all of them and some may be forgotten over time, until they resurface like this.

**Reply by setanarut** [reply]
95% is fine. As you go below about 90% the interface will become a nightmare.

**Reply by Calinou** [reply]
@setanarut I have to remind you that we have a [Code of Conduct](https://godotengine.org/code-of-conduct). Please stay constructive.
